### PR TITLE
Implement 'peer' sync mode

### PR DESCRIPTION
**Pull request by platfowner**
Change summary:
- Implement 'peer' sync mode
  - Add SYNC_MODE='peer' option 
  - Add SNAPSHOT_CHUNK_REQUEST and SNAPSHOT_CHUNK_RESPONSE message types
  - Add STATE_LOADING / STATE_SYNCING / READY_TO_START node states
- Replace getNumBlockFiles() with getLatestBlockInfo()
- Set default SYNC_MODE='fast'

Related issues:
- https://github.com/ainblockchain/ain-blockchain/issues/713

### CODE REVIEW

File: common/file-util.js [hunk 1]
```diff
@@ -60,28 +60,81 @@ class FileUtil {
...
+    } catch (err) {
+      logger.debug(
+          `[${LOG_HEADER}] Failed to read block files from ${blockFilePathPrefix}: ${err.stack}`);
+      return { latestBlockPath, latestBlockNumber };
+    }
+    if (fileList.length === 0) {
+      return { latestBlockPath, latestBlockNumber };
+    }
+    for (const file of fileList) {
+      const numString = _.get(_.split(file, `.${JSON_GZIP_FILE_EXTENSION}`), 0);
```
Comment (liayoo) [inline common/file-util.js:127]: how about defining a util function for L127-129 (e.g. getBlockNumberFromGzipFile()) and use it here & in getLatestSnapshotInfo()?

File: common/file-util.js [hunk 2]
```diff
@@ -127,25 +180,58 @@ class FileUtil {
     return _.endsWith(filePath, '.gz');
   }
 
+  static async processChunkedJsonAsync(filePath, chunkCallback, endCallback) {
+    const LOG_HEADER = 'readChunkedJsonAsyncWithCallbacks';
```
Comment (liayoo) [inline common/file-util.js:184]: processChunkedJsonAsync?

File: node/index.js [hunk 3]
```diff
@@ -221,63 +225,118 @@ class BlockchainNode {
...
+      logger.info(`[${LOG_HEADER}] Now node in READY_TO_START state!`);
     }
+  }
+
+  async loadLatestSnapshot() {
+    const LOG_HEADER = 'loadLatestSnapshot';
+
+    const latestSnapshotInfo = FileUtil.getLatestSnapshotInfo(this.snapshotDir);
+    const latestSnapshotPath = latestSnapshotInfo.latestSnapshotPath;
+    if (latestSnapshotPath) {
```
Comment (minsulee2) [inline node/index.js:263]: What happens if latestSnapshotPath is not set? Just return true and it assumes this has nothing in snapshot?

File: node/index.js [hunk 4]
```diff
@@ -672,11 +732,17 @@ class BlockchainNode {
...
+      logger.error(`[${LOG_HEADER}] Invalid latest block number: ${latestBlockNumber}`);
+      return false;
+    }
+    const fromBlockNumber = (NodeConfigs.SYNC_MODE === SyncModeOptions.FAST ||
+        NodeConfigs.SYNC_MODE === SyncModeOptions.PEER) ?
+        Math.max(latestSnapshotBlockNumber, 0) : 0;
     let nextBlock = null;
     let proposalTx = null;
-    for (let number = fromBlockNumber; number < numBlockFiles; number++) {
+    for (let number = fromBlockNumber; number <= latestBlockNumber; number++) {
```
Comment (minsulee2) [inline node/index.js:745]: Oh, Nice updates!!

File: node/index.js [hunk 5]
```diff
@@ -221,63 +225,118 @@ class BlockchainNode {
...
+  }
+
+  resetLatestSnapshot() {
+    this.setLatestSnapshot(null, null);
+  }
 
-    let latestSnapshot = null;
-    let latestSnapshotPath = null;
-    let latestSnapshotBlockNumber = -1;
+  checkSyncMode() {
```
Comment (liayoo) [inline node/index.js:240]: how about setSyncMode() or setNodeState()?

File: p2p/index.js [hunk 6]
```diff
@@ -732,16 +851,62 @@ class P2pClient {
...
+    logger.info(`[${LOG_HEADER}] Handling a snapshot chunk of chunkIndex ${chunkIndex} ` +
+        `and numChunks ${numChunks}.`);
+    const chunkArraySize = _.get(this.stateSyncInProgress, 'chunks.length', null);
+    if (numChunks > 0) {
+      if (numChunks !== chunkArraySize) {
+        logger.error(`[${LOG_HEADER}] Mismatched numChunks: ${numChunks} !== ${chunkArraySize}`);
+        return;
+      }
+      await this.buildSnapshotAndStartBlockchainNode();
+    } else if (chunkIndex >= 0) {
```
Comment (minsulee2) [inline p2p/index.js:873]: What about two different if conditions if these are checking different conditions?

File: p2p/index.js [hunk 7]
```diff
@@ -732,16 +851,62 @@ class P2pClient {
...
+      this.updateStateSyncPeer(chunk, chunkIndex);
+    }
+  }
+
+  async buildSnapshotAndStartBlockchainNode() {
+    const LOG_HEADER = 'buildSnapshotAndStartBlockchainNode';
+
+    const snapshot = FileUtil.buildObjectFromChunks(this.stateSyncInProgress.chunks);
+    const source = `${this.stateSyncInProgress.address} (${this.stateSyncInProgress.p2pUrl})`;
+    const blockNumber =
```
Comment (minsulee2) [inline p2p/index.js:887]: Can this be in one line?

File: node/index.js [hunk 8]
```diff
@@ -221,63 +225,118 @@ class BlockchainNode {
...
+    // 4. Execute transactions from the pool.
     logger.info(`[${LOG_HEADER}] Executing the transaction from the tx pool..`);
     this.db.executeTransactionList(
         this.tp.getValidTransactions(null, this.stateManager.getFinalVersion()), false, true,
         this.bc.lastBlockNumber() + 1, this.bc.lastBlockTimestamp());
 
-    // 6. Node status changed: STARTING -> SYNCING.
-    this.state = BlockchainNodeStates.SYNCING;
-    logger.info(`[${LOG_HEADER}] Now node in SYNCING state!`);
+    // 5. Node status changed: STATE_LOADING/STATE_SYNCING -> CHAIN_SYNCING.
```
Comment (liayoo) [inline node/index.js:334]: question) Did I understand correctly that it should be READY_TO_START state at this point? (https://github.com/ainblockchain/ain-blockchain/pull/963/files#diff-a6f7d482db046ebe17836e96f3280b27890bba7136c1269fe32e4a73d4f7b345R452)

File: p2p/index.js [hunk 9]
```diff
@@ -383,38 +423,62 @@ class P2pClient {
...
+    const LOG_HEADER = 'prepareBlockchainNode';
 
-    if (numLivePeers === 0) {
-      logger.info(`[${LOG_HEADER}] Starting blockchain node without peers..`);
-      if (!(await this.server.node.initNode(true))) {
+    logger.info(
+        `[${LOG_HEADER}] Preparing blockchain node with isFirstNode = ${this.isFirstNode} ..`);
+    this.server.node.checkSyncMode();
+    if (this.server.node.state === BlockchainNodeStates.STATE_SYNCING) {
+      // Wait until some peer connections are made.
```
Comment (liayoo) [inline p2p/index.js:433]: could we use # outbound peers to check this?

File: p2p/index.js [hunk 10]
```diff
@@ -732,16 +851,62 @@ class P2pClient {
...
+        return;
+      }
+      this.updateStateSyncPeer(chunk, chunkIndex);
+    }
+  }
+
+  async buildSnapshotAndStartBlockchainNode() {
+    const LOG_HEADER = 'buildSnapshotAndStartBlockchainNode';
+
+    const snapshot = FileUtil.buildObjectFromChunks(this.stateSyncInProgress.chunks);
```
Comment (liayoo) [inline p2p/index.js:885]: should we add a null check here?

File: common/file-util.js [hunk 11]
```diff
@@ -127,25 +180,58 @@ class FileUtil {
...
+    try {
+      return new Promise((resolve) => {
+        const transformStream = JSONStream.parse('docs.*');
+        let numChunks = 0;
+        fs.createReadStream(filePath)
+          .pipe(zlib.createGunzip())
+          .pipe(transformStream)
+          .on('data', (data) => {
```
Comment (minsulee2) [inline common/file-util.js:192]: Ohhhh I love it. Question) Does this call all data chunk before getting into the 'end' stage?

File: p2p/index.js [hunk 12]
```diff
@@ -874,6 +1040,31 @@ class P2pClient {
...
+      p2pUrl,
+      chunks: [],
+      lastChunkIndex: -1,
+      updatedAt: Date.now
+    };
+  }
+
+  updateStateSyncPeer(chunk, lastChunkIndex) {
+    if (!this.stateSyncInProgress) return;
+    this.stateSyncInProgress.chunks.push(chunk);
```
Comment (liayoo) [inline p2p/index.js:1059]: question) can we assume that the order of chunks received will maintain the original order?

File: common/file-util.js [hunk 13]
```diff
@@ -60,28 +60,81 @@ class FileUtil {
...
+
+    let latestBlockPath = null;
+    let latestBlockNumber = -1;
+
+    const blockDirPathPrefix = path.join(chainPath, BlockchainConsts.CHAINS_N2B_DIR_NAME);
+    let dirList = [];
+    try {
+      dirList = fs.readdirSync(blockDirPathPrefix);
+    } catch (err) {
+      logger.debug(
```
[the remark this answers is not in the document]
Comment (platfowner) [reply common/file-util.js]: Done, thanks!

File: node/index.js (hunk 3 above)
Comment (platfowner) [reply node/index.js:263]: That's also considered as a normal case, i.e., there is not snapshot file created yet. So the node just starts without snapshot files.

File: p2p/index.js (hunk 6 above)
Comment (platfowner) [reply p2p/index.js]: Tried to simplify it. PTAL

File: p2p/index.js (hunk 7 above)
Comment (platfowner) [reply p2p/index.js]: Done, thanks!

File: common/file-util.js (hunk 11 above)
Comment (platfowner) [reply common/file-util.js:194]: Yes, that's how the JSONStream.parse() works.

File: common/file-util.js [hunk 14]
```diff
@@ -60,28 +60,81 @@ class FileUtil {
...
+      bNum = CommonUtil.isNumber(bNum) ? bNum : -1;
+      return bNum - aNum;
+    })[0];
+
+    const blockFilePathPrefix = path.join(blockDirPathPrefix, latestBlockDir);
+    let fileList = [];
+    try {
+      fileList = fs.readdirSync(blockFilePathPrefix);
+    } catch (err) {
+      logger.debug(
```
[the remark this answers is not in the document]
Comment (platfowner) [reply common/file-util.js]: Done, thanks!

File: common/file-util.js (hunk 1 above)
Comment (platfowner) [reply common/file-util.js]: Done, thanks!

File: common/file-util.js (hunk 2 above)
Comment (platfowner) [reply common/file-util.js]: Just fixed, thanks!

File: node/index.js (hunk 5 above)
Comment (platfowner) [reply node/index.js]: Tried to change it to setNodeStateBySyncMode(). PTAL

File: node/index.js (hunk 8 above)
Comment (platfowner) [reply node/index.js]: Just fixed, thanks!

File: p2p/index.js (hunk 9 above)
Comment (platfowner) [reply p2p/index.js]: Tried to replace the time buffer approach with a retrial approach (in setTimerForPeerAddressResponse()) with the outbound size checking logic. PTAL

File: p2p/index.js (hunk 10 above)
Comment (platfowner) [reply p2p/index.js:889]: Done, thanks!

File: p2p/index.js (hunk 12 above)
Comment (platfowner) [reply p2p/index.js:1068]: It simply checks the order of the chunks in handleSnapshotChunk(). I think this makes sense as the chunks are delivered by streaming through a p2p connection where the order of the chunks are stable?

**Comment by platfowner** [comment]
Thanks for the comments and questions. 
They were all addressed. PTAL @minsulee2 @liayoo

File: p2p/index.js [hunk 15]
```diff
@@ -732,16 +850,66 @@ class P2pClient {
...
+      this.updateStateSyncStatus(chunk, chunkIndex);
+    }
+  }
+
+  async buildSnapshotAndStartBlockchainNode() {
+    const LOG_HEADER = 'buildSnapshotAndStartBlockchainNode';
+
+    const chunks = _.get(this.stateSyncInProgress, 'chunks', null);
+    if (!chunks || chunks.length === 0) {
+      logger.error(`[${LOG_HEADER}] Invalid chunks: ${JSON.stringify(chunks)}`);
```
Comment (liayoo) [inline p2p/index.js:886]: Should we re-assign a peer and retry? Maybe leave a todo for now?

File: p2p/index.js (hunk 12 above)
Comment (liayoo) [reply p2p/index.js:1068]: I see, I wasn't sure if the sending order is preserved on the receiving side. Thanks for the explanation!

**Comment by platfowner** [comment]
Comments addressed, PTAL~ @liayoo @minsulee2

File: p2p/index.js [hunk 16]
```diff
@@ -732,16 +850,66 @@ class P2pClient {
...
+      return;
+    }
+    const snapshot = FileUtil.buildObjectFromChunks(chunks);
```
[the remark this answers is not in the document]
Comment (platfowner) [reply p2p/index.js:889]: Ah, it seems that github has some bug that sometimes I cannot add any comments to existing thread. So I just tried to add a new comment. The changes were:
```
const snapshot = FileUtil.buildObjectFromChunks(this.stateSyncInProgress.chunks);
```
=>
```
const snapshot = FileUtil.buildObjectFromChunks(chunks);
```

File: p2p/index.js (hunk 15 above)
Comment (platfowner) [reply p2p/index.js]: To be consistent with other cases, just tried to stop the blockchain. PTAL~

File: p2p/index.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (liayoo) [reply p2p/index.js:897]: Ahhh gotcha. Looks good! Thanks for the updates!

File: p2p/index.js [hunk 17]
```diff
@@ -732,16 +852,72 @@ class P2pClient {
...
+      logger.error(`[${LOG_HEADER}] Mismatched senderAddress: ${senderAddress} !== ${peerAddress}`);
+      return;
+    }
+    logger.info(`[${LOG_HEADER}] Handling a snapshot chunk of chunkIndex ${chunkIndex} ` +
+        `and numChunks ${numChunks}.`);
+    const chunkArraySize = _.get(this.stateSyncInProgress, 'chunks.length', null);
+    if (chunkIndex === -1) {  // 'end' message
+      if (numChunks !== chunkArraySize) {
+        logger.error(`[${LOG_HEADER}] Mismatched numChunks: ${numChunks} !== ${chunkArraySize}`);
+        this.server.node.state = BlockchainNodeStates.STOPPED;
```
Comment (liayoo) [inline p2p/index.js:871]: checked ✔️

File: p2p/index.js [hunk 18]
```diff
@@ -732,16 +852,72 @@ class P2pClient {
...
+        logger.error(`[${LOG_HEADER}] Blockchain node stopped!`);
+        return;
+      }
+      await this.buildSnapshotAndStartBlockchainNode();
+    } else {  // 'data' message
+      if (chunkIndex !== chunkArraySize) {
+        logger.error(`[${LOG_HEADER}] Mismatched chunkIndex: ${chunkIndex} !== ${chunkArraySize}`);
+        this.server.node.state = BlockchainNodeStates.STOPPED;
```
Comment (liayoo) [inline p2p/index.js:879]: ditto ✔️

**Review by liayoo — APPROVED** [review]
LGTM, thanks for the nice feature!

**Comment by platfowner** [comment]
Thanks for the review! Now merging..

**Pull request merged**